File: v3/screenshots/cbt_api_screenshots.php
<?php

/* CrossBrowserTesting.com API v3 Screenshots sample script 
* - requires PHP >= 5.2
* - run on the command line
* - requires libcurl package to be installed
*/

//set authentication info
$username = ""; //email address for your account
$password = ""; //authkey for your account

define ('EOL', "\n");
define ('TAB', "\t");


class ScreenshotTestApi{

	public $baseUrl = "https://crossbrowsertesting.com/api/v3/screenshots";

	public $currentTest = NULL;
	public $allTests = array();
	public $recordCout = 0;

	function __construct($username, $password, $screenshot_test_id=NULL) {
		$this->user = $username;
		$this->pass = $password;

		//if test id provided initialize this test to that test data
        if ($screenshot_test_id){
            $url = $this->baseUrl . "/" . $screenshot_test_id;
            $this->currentTest = $this->callApi($url);
        }
	}

	function startNewTest($params){
		$this->currentTest = $this->callApi($this->baseUrl, 'POST', $params);
	}

	function updateTestInfo(){
		$url = $this->baseUrl . "/" . $this->getTestId();
		return $this->callApi($url, 'GET');
	}

	function getTestId(){
		return $this->currentTest->screenshot_test_id;
	}

	function printTestBrowsers(){
		if ($this->currentTest){
			foreach ($this->currentTest->versions[0]->results as $result) {
				print $result->os->name  . TAB . $result->browser->name . TAB . $result->resolution->name . EOL;
			}
		}
	}

	function isTestComplete(){
		$this->currentTest = $this->updateTestInfo();
		return !$this->currentTest->versions[0]->active;
	}

	function getScreenshotBrowsers(){
		$url = $this->baseUrl . "/browsers";
		return $this->callApi($url, 'GET');
	}

	function getAllTests($params = false){
		$url = $this->baseUrl;
		$result = $this->callApi($url, 'GET',$params);
		$this->recordCount = $result->meta->record_count;
        $this->allTests = $result->screenshots;
        return $this->allTests;
	}

	function callApi($api_url, $method = 'GET', $params = false){
		$apiResult = NULL;

	    $process = curl_init();

	    switch ($method){
	        case "POST":
	            curl_setopt($process, CURLOPT_POST, 1);

	            if ($params){
	                curl_setopt($process, CURLOPT_POSTFIELDS, http_build_query($params));
	                curl_setopt($process, CURLOPT_HTTPHEADER, array('User-Agent: php')); //important
	            }
	            break;
	        case "PUT":
	            curl_setopt($process, CURLOPT_CUSTOMREQUEST, "PUT");
	            if ($params){
	                curl_setopt($process, CURLOPT_POSTFIELDS, http_build_query($params));
	                curl_setopt($process, CURLOPT_HTTPHEADER, array('User-Agent: php')); //important
	            }
	            break;
	         case 'DELETE':
	         	curl_setopt($process, CURLOPT_CUSTOMREQUEST, "DELETE");
	         	break;
	        default:
	            if ($params){
	                $api_url = sprintf("%s?%s", $api_url, http_build_query($params));
	            }
	    }

	    // Optional Authentication:
	    curl_setopt($process, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
	    curl_setopt($process, CURLOPT_USERPWD, $this->user . ":" . $this->pass);

	    curl_setopt($process, CURLOPT_URL, $api_url);
	    curl_setopt($process, CURLOPT_RETURNTRANSFER, 1);
	    curl_setopt($process, CURLOPT_TIMEOUT, 30);

	    $apiResult->content = curl_exec($process);
		$apiResult->httpResponse = curl_getinfo($process);	
		$apiResult->errorMessage =  curl_error($process);
		$apiResult->params = $params;

		curl_close($process);

		//print_r($apiResult);

		$paramsString = $params ? http_build_query($params) : '';
		$response = json_decode($apiResult->content);

		if ($apiResult->httpResponse['http_code'] != 200){
			$message = 'Error calling "' . $apiResult->httpResponse['url'] . '" ';
			$message .= (isset($paramsString) ? 'with params "'.$paramsString.'" ' : ' ');
			$message .= '. Returned HTTP status ' . $apiResult->httpResponse['http_code'] . ' ';
			$message .= (isset($apiResult->errorMessage) ? $apiResult->errorMessage : ' ');
			$message .= (isset($response->message) ? $response->message : ' ');
			die($message);
		}
		else {
			$response = json_decode($apiResult->content);
			if (isset($response->status)){
				die('Error calling "' . $apiResult->httpResponse['url'] . '"' .(isset($paramsString) ? 'with params "'.$paramsString.'"' : '') . '". ' . $response->message);			
			}
		}

	    return $response;
	}
}



function runNewTest(){
	global $username, $password;

	print EOL.'** Starting CrossBrowserTesting.com API v3 Run Screenshot Test example **'.EOL;

	//set parameters we want for the screenshot test
    $params["url"] = "http://www.google.com";

    //set browsers
    $params["browsers"] = array();
    $params["browsers"][] = "Win7x64|IE11|1400x1050";
    $params["browsers"][] = "Mac10.9|chrome-latest";
    $params["browsers"][] = "Win10|ff-latest";

	$screenshot = new ScreenshotTestApi($username,$password);

	print "starting new screenshot test for " . $params["url"].EOL;
    $screenshot->startNewTest($params);

    print "screenshot_test_id is " . $screenshot->getTestId().EOL;
    print "view Screenshot Test on web here: https://app.crossbrowsertesting.com/screenshots/" . $screenshot->getTestId().EOL;

    print EOL."browsers to be tested are: ".EOL;
    $screenshot->printTestBrowsers();

    print "waiting on test to complete".EOL;
    $tries = 0;
    $maxTries = 100;
    while ($tries < $maxTries){
        
        if ($screenshot->isTestComplete()){
            print "screenshot test complete".EOL;
            break;
        }
        else{
            sleep(2);
            $tries += 1;
        }
	}
    if ($tries >= $maxTries){
        die("screenshot did not complete after " . str($tries*2) . " seconds!".EOL);
    }
}

function viewTestHistory(){
	global $username, $password;

	print EOL."** Starting CrossBrowserTesting.com API v3 View Screenshot History example **".EOL;

	//set paging options
	$params["start"] = 0; //start with the last test run
    $params["num"] = 20; //how many to retrieve

    //filter results
    $params["url"] = "google"; //filter for only tests run that have 'google' somewhere in the URL
    $params["start_date"] = "2014-06-01"; //fitler to only tests run within a date range
    $params["end_date"] = "2014-10-01";
    $params["archived"] = false; //only include screenshot tests that are not archived

    //create api object and set auth info
    print "retrieving test history".EOL;
    $screenshots = new ScreenshotTestApi($username,$password);
    $allTests = $screenshots->getAllTests($params);

    //show total number of tests
    print "There are " . $screenshots->recordCount . " tests for  the filters provided, showing " . count($allTests) . EOL;

    //print out results
    for ($i=0; $i<count($allTests); $i++){
        $test = $allTests[$i];
        $version = $test->versions[0];
        $start_date = $version->start_date;
        print ($i+1) . TAB . $start_date . TAB . $test->screenshot_test_id  . TAB . $test->url.EOL;
    }

}

function listScreenshotBrowsers(){
    print EOL."** Starting CrossBrowserTesting.com API v3 List Screenshot Browsers example **".EOL;

    //create api object and set auth info
    $screenshotApi = new ScreenshotTestApi($username,$password);
    $oss = $screenshotApi->getScreenshotBrowsers();

    foreach ($oss as $os){
        foreach($os->browsers as $browser){
            print $os->name . TAB . $browser->name . TAB . $os->api_name . "|" . $browser->api_name.EOL;
        }
    }
}

// uncomment a function to run
// listScreenshotBrowsers();
// runNewTest();
// viewTestHistory();

?>

File: v3/screenshots/cbt_api_screenshots.py

"""
CrossBrowserTesting.com API v3 Run Screenshots sample script 

1. Install requests module: http://docs.python-requests.org/en/latest/user/install/#install
2. Run on Python 2.7 or greater
3. Run on Python Interpreter (command line)

"""
import requests, json, time

#import httplib as http_client
#http_client.HTTPConnection.debuglevel = 1

#set authentication info
username = "" #email address for your account
password = "" #authkey for your account

EOL = "\n"
TAB = "\t"

class ScreenshotTestApi:

    def __init__(self, username, password, screenshot_test_id=None):

        self.auth = (username, password)
        self.baseUrl = "https://crossbrowsertesting.com/api/v3/screenshots"
        self.currentTest = None
        self.allTests = []
        self.recordCount = 0

        #if test id provided initialize this test to that test data
        if screenshot_test_id:
            url = self.baseUrl + "/" + str(screenshot_test_id)
            self.currentTest = self.callApi(url)
    
    def startNewTest(self, params):
         self.currentTest = self.callApi(self.baseUrl, "POST", params)

    def updateTestInfo(self):
        url = self.baseUrl + "/" + str(self.getTestId())
        return self.callApi(url)

    def getTestId(self):
        return self.currentTest["screenshot_test_id"]

    def printTestBrowsers(self):
        if (self.currentTest):
            for result in self.currentTest["versions"][0]["results"]:
                print result["os"]["name"] + TAB + result["browser"]["name"] + TAB + result["resolution"]["name"]

    def isTestComplete(self):
        self.currentTest = self.updateTestInfo()
        return not self.currentTest["versions"][0]["active"]

    def getScreenshotBrowsers(self):
        url = self.baseUrl + "/browsers"
        return self.callApi(url)

    def getAllTests(self, params={}):
        result = self.callApi(self.baseUrl,"GET", params)
        self.recordCount = result["meta"]["record_count"]
        self.allTests = result["screenshots"]
        return self.allTests

    def callApi(self, url, method="GET", params={}):

        if method == "GET":
            response = requests.get(url,auth=self.auth,params=params)
        elif method == "POST":
            response = requests.post(url,auth=self.auth,params=params)
        elif method == "PUT":
            response = requests.put(url,auth=self.auth,params=params)
        elif method == "DELETE":
            response = requests.delete(url,auth=self.auth,params=params)
        else:
            response = requests.get(url,auth=self.auth,params=params)

        print response.text
        result = json.loads(response.text)

        if "status" in result:
            print "http error " + str(result["status"]),
            if "message" in result:
                print " : " + result["message"]
            quit()
        else:
            return result
    

def runNewTestAndWait():

    print EOL + "** Starting CrossBrowserTesting.com API v3 Run Screenshot Test example **"

    #set parameters we want for the screenshot test
    params = {}
    #params["url"] = "http://www.google.com"
    params["url"] = "http://www.midsouthproducts.com/collections/elvis-presley/products/frame-elvis-dancing-pewter"

    #set browsers
    params["browsers"] = []
    #params["browsers"].append("Win7x64|IE10|1400x1050")
    #params["browsers"].append("Win7x64|chrome-latest|1400x1050")
    #params["browsers"].append("Win7x64|ff-latest|1400x1050")
    #params["browsers"].append("Mac10.9|chrome-latest-1")

    #other options
    params["browser_list_name"] = "Popular Browsers"
    params["login"] = "www.midsouthproducts.com login"
    ##params["login"] = "mydomain.com login" #valid only if you've created a login profile with this name
    ##params["basic_username"] = "username" #for basic auth urls only
    ##params["basic_password"] = "password" #for basic auth urls only
    params["delay"] = 4 #delay for number of seconds to wait after page is loaded to start capturing screenshots

    #create api object and set auth info
    screenshot = ScreenshotTestApi(username,password)

    print "starting new screenshot test for " + params["url"]
    screenshot.startNewTest(params)

    print "screenshot_test_id is " + str(screenshot.getTestId())
    print "view Screenshot Test on web here: https://app.crossbrowsertesting.com/screenshots/" + str(screenshot.getTestId())

    print EOL + "browsers to be tested are: "
    screenshot.printTestBrowsers()

    print EOL + "waiting on test to complete"
    tries = 0
    maxTries = 100
    while tries < maxTries:
        
        if screenshot.isTestComplete():
            print "screenshot test complete"
            break
        else:
            time.sleep(2)
            tries += 1

    if tries >= maxTries:
        print "screenshot did not complete after " + str(tries*2) + " seconds!"
        quit()

def runNewTest():

    print EOL + "** Starting CrossBrowserTesting.com API v3 Run Screenshot Test example **"

    #set parameters we want for the screenshot test
    params = {}
    #params["url"] = "http://212.121.155.205/WFRM-QA-Support/CBTWebPage/600382124/WOFF2.html"
    params["url"] = "http://www.google.com"
    #params["url"] = "http://www.midsouthproducts.com/collections/elvis-presley/products/frame-elvis-dancing-pewter"

    #set browsers
    # params["browsers"] = []
    # params["browsers"].append("Win8.1|chrome-latest|2560x1600")
    # params["browsers"].append("Win8.1|chrome-latest|1920x1080")
    #params["browsers"].append("Win7x64|ff-latest|1400x1050")
    #params["browsers"].append("Win7x64|ff-latest|1400x1050")
    #params["browsers"].append("Mac10.9|chrome-latest")

    #other options
    params["browser_list_name"] = "simple-list"
    #params["login"] = "www.midsouthproducts.com login"
    ##params["login"] = "mydomain.com login" #valid only if you've created a login profile with this name
    ##params["basic_username"] = "username" #for basic auth urls only
    ##params["basic_password"] = "password" #for basic auth urls only
    #params["delay"] = 4 #delay for number of seconds to wait after page is loaded to start capturing screenshots

    #create api object and set auth info
    screenshot = ScreenshotTestApi(username,password)

    print "starting new screenshot test for " + params["url"]
    screenshot.startNewTest(params)

    print "screenshot_test_id is " + str(screenshot.getTestId())
    print "view Screenshot Test on web here: https://app.crossbrowsertesting.com/screenshots/" + str(screenshot.getTestId())

    print EOL + "browsers to be tested are: "
    screenshot.printTestBrowsers()

def viewTestHistory():

    print EOL + "** Starting CrossBrowserTesting.com API v3 View Screenshot History example **"

    #set paging options
    params = {}
    params["start"] = 0 #start with the last test run
    params["num"] = 20 #how many to retrieve

    #filter results
    params["url"] = "google" #filter for only tests run that have 'google' somewhere in the URL
    params["start_date"] = "2014-06-01" #fitler to only tests run within a date range
    params["end_date"] = "2014-10-01"
    params["archived"] = False #only include screenshot tests that are not archived

    #create api object and set auth info
    print "retrieving test history"
    screenshots = ScreenshotTestApi(username,password)
    allTests = screenshots.getAllTests(params)

    #show total number of tests
    print "There are " + str(screenshots.recordCount) + " tests for  the filters provided, showing " + str(len(allTests))

    for i in range(len(allTests)):
        test = allTests[i]
        version = test["versions"][0]
        start_date = version["start_date"]
        print str(i+1) + TAB + start_date + TAB + str(test["screenshot_test_id"]) + TAB + test["url"]

def listScreenshotBrowsers():
    print EOL + "** Starting CrossBrowserTesting.com API v3 List Screenshot Browsers example **"

    #create api object and set auth info
    screenshotApi = ScreenshotTestApi(username,password)
    oss = screenshotApi.getScreenshotBrowsers()

    for os in oss:
        for browser in os["browsers"]:
            print os["name"] + TAB + browser["name"] + TAB + os["api_name"] + "|" + browser["api_name"]

# uncomment a function to run
# listScreenshotBrowsers()
# runNewTest()
# viewTestHistory()